src/Microsoft.AspNetCore.Server.IISIntegration/Server/IISHttpContext.ReadWrite.cs

```diff
@@ -0,0 +1,431 @@
+
+                            if (!buffer.IsEmpty)
+                            {
+                                // Write to IIS buffers
+                                // Guaranteed to write the entire buffer to IIS
+                                await WriteToIISAsync(buffer);
+                            }
+                            else if (readResult.IsCompleted)
+                            {
+                                _doneWriting = true;
+                                break;
+                            }
+                            else
+                            {
+                                // Flush of zero bytes
+                                await FlushToIISAsync();
+                            }
+                        }
+                        finally
+                        {
+                            // Always Advance the data pointer to the end of the buffer.
+                            Output.Reader.AdvanceTo(buffer.End);
+                        }
+                    }
+
+                    // Check if there was an upgrade. If there is, we will replace the request and response bodies with
+                    // two seperate loops. These will still be using the same Input and Output pipes here.
+                    if (_upgradeTcs?.TrySetResult(null) == true)
+                    {
+                        await StartBidirectionalStream();
+
+                        // Input and Output will be closed in StartBidirectionalStream.
+                        // We can return at this point.
+                        return;
+                    }
+
+                    // Now we handle the read. 
+                    var memory = Input.Writer.GetMemory();
+                    _inputHandle = memory.Retain(true);
+
+                    try
+                    {
+                        // Lock around invoking ReadFromIISAsync as we don't want to call CancelIo
+                        // when calling read
+                        var read = await ReadFromIISAsync(memory.Length);
+
+                        // read value of 0 == done reading
+                        // read value of -1 == read cancelled, continue trying to read
+                        // but don't advance the pipe
+                        if (read == 0)
+                        {
+                            _doneReading = true;
+                        }
+                        else if (read != -1)
+                        {
+                            Input.Writer.Advance(read);
+                        }
+                    }
+                    finally
+                    {
+                        // Always commit any changes to the Input pipe
+                        Input.Writer.Commit();
+                        _inputHandle.Dispose();
+                    }
+
+                    // Flush the read data for the Input Pipe writer
+                    var flushResult = await Input.Writer.FlushAsync();
```

Comment:
You're going to wake the reader up even if you read nothing or you cancelled. Is that intentional?

Reply:
It wasn't. I added a continue before.